File: widgets/track_card.py
from PySide6.QtWidgets import (
    QFrame, QHBoxLayout, QVBoxLayout, QLabel, QSlider,
    QComboBox, QPushButton, QLineEdit, QSizePolicy
)
from PySide6.QtCore import Qt, Signal, QTimer
from PySide6.QtGui import QFont, QColor
from widgets.knob import CustomKnob
from widgets.level_meter import LevelMeter

class TrackCard(QFrame):
    """Card-style Track widget displaying track settings and VU meter."""
    trackSelected = Signal(object)  # Emitted when the card is clicked/selected
    trackRemoved = Signal(object)   # Emitted when delete button is pressed
    
    def __init__(self, track, audio_engine, parent=None):
        super().__init__(parent)
        self.track = track
        self.audio_engine = audio_engine
        self.is_selected = False
        
        self.setFrameShape(QFrame.Shape.StyledPanel)
        self.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Preferred)
        self.setMinimumHeight(150)
        self.setObjectName("TrackCard")
        
        self.setup_ui()
        self.update_selection_style()
        
        # Setup level meter polling timer
        self.poll_timer = QTimer(self)
        self.poll_timer.timeout.connect(self.update_levels)
        self.poll_timer.start(33)  # Polling level at ~30 FPS
        
    def setup_ui(self):
        # Master horizontal layout for card contents
        main_layout = QHBoxLayout(self)
        main_layout.setContentsMargins(10, 8, 10, 8)
        main_layout.setSpacing(10)
        
        # --- PANEL 1: Track Details & State Buttons ---
        details_layout = QVBoxLayout()
        details_layout.setSpacing(6)
        
        # Editable Track Name
        self.name_edit = QLineEdit(self.track.name)
        self.name_edit.setObjectName("TrackNameEdit")
        self.name_edit.setToolTip("Double-click to rename track")
        self.name_edit.editingFinished.connect(self.on_rename)
        details_layout.addWidget(self.name_edit)
        
        # State Buttons Row: Mute [M], Solo [S], Arm [R]
        btn_layout = QHBoxLayout()
        btn_layout.setSpacing(4)
        
        self.btn_mute = QPushButton("M")
        self.btn_mute.setCheckable(True)
        self.btn_mute.setChecked(self.track.mute)
        self.btn_mute.setObjectName("MuteButton")
        self.btn_mute.setToolTip("Mute Track")
        self.btn_mute.clicked.connect(self.on_mute_clicked)
        
        self.btn_solo = QPushButton("S")
        self.btn_solo.setCheckable(True)
        self.btn_solo.setChecked(self.track.solo)
        self.btn_solo.setObjectName("SoloButton")
        self.btn_solo.setToolTip("Solo Track")
        self.btn_solo.clicked.connect(self.on_solo_clicked)
        
        self.btn_arm = QPushButton("R")
        self.btn_arm.setCheckable(True)
        self.btn_arm.setChecked(self.track.armed)
        self.btn_arm.setObjectName("ArmButton")
        self.btn_arm.setToolTip("Arm Track for Input Monitoring")
        self.btn_arm.clicked.connect(self.on_arm_clicked)
        
        btn_layout.addWidget(self.btn_mute)
        btn_layout.addWidget(self.btn_solo)
        btn_layout.addWidget(self.btn_arm)
        details_layout.addLayout(btn_layout)
        
        # Input Channel Selector
        self.combo_input = QComboBox()
        self.combo_input.setObjectName("InputChannelCombo")
        self.combo_input.setToolTip("Select Track Input Channel")
        self.populate_inputs()
        self.combo_input.currentIndexChanged.connect(self.on_input_changed)
        details_layout.addWidget(self.combo_input)
        
        # Delete track button
        self.btn_delete = QPushButton("Delete Track")
        self.btn_delete.setObjectName("DeleteTrackButton")
        self.btn_delete.clicked.connect(self.on_delete_clicked)
        details_layout.addWidget(self.btn_delete)
        
        main_layout.addLayout(details_layout)
        
        # --- PANEL 2: Panning Knob & FX selector ---
        panning_layout = QVBoxLayout()
        panning_layout.setAlignment(Qt.AlignmentFlag.AlignCenter)
        
        # Custom Pan Knob
        self.pan_knob = CustomKnob(
            label="PAN",
            min_val=-1.0,
            max_val=1.0,
            default_val=self.track.pan,
            decimals=2
        )
        self.pan_knob.valueChanged.connect(self.on_pan_changed)
        panning_layout.addWidget(self.pan_knob)
        
        # FX Select Button
        self.btn_fx = QPushButton("VST / FX")
        self.btn_fx.setObjectName("FxButton")
        self.btn_fx.setToolTip("View/edit effects rack for this track")
        self.btn_fx.clicked.connect(self.on_fx_clicked)
        panning_layout.addWidget(self.btn_fx)
        
        main_layout.addLayout(panning_layout)
        
        # --- PANEL 3: Volume Fader ---
        vol_layout = QVBoxLayout()
        vol_layout.setSpacing(4)
        vol_layout.setAlignment(Qt.AlignmentFlag.AlignCenter)
        
        self.vol_slider = QSlider(Qt.Orientation.Vertical)
        self.vol_slider.setMinimum(-600)  # -60.0 dB
        self.vol_slider.setMaximum(60)    # +6.0 dB
        self.vol_slider.setValue(int(self.track.volume * 10))
        self.vol_slider.setObjectName("VolumeSlider")
        self.vol_slider.valueChanged.connect(self.on_volume_changed)
        vol_layout.addWidget(self.vol_slider)
        
        self.vol_label = QLabel("0.0 dB")
        self.vol_label.setObjectName("VolDbLabel")
        self.vol_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.vol_label.setFont(QFont("Consolas", 8))
        vol_layout.addWidget(self.vol_label)
        self.update_volume_label(self.track.volume)
        
        main_layout.addLayout(vol_layout)
        
        # --- PANEL 4: Custom vertical LED VU Meter ---
        self.level_meter = LevelMeter()
        main_layout.addWidget(self.level_meter)
        
        # Apply CSS stylesheets directly (QSS)
        self.setStyleSheet("""
            TrackCard {
                background-color: #2a2a2a;
                border: 1px solid #333333;
                border-left: 4px solid transparent;
                border-radius: 4px;
            }
            QLineEdit#TrackNameEdit {
                background: transparent;
                color: #ffffff;
                border: none;
                border-bottom: 1px solid transparent;
                font-family: "Segoe UI", sans-serif;
                font-size: 13px;
                font-weight: bold;
                padding-bottom: 2px;
            }
            QLineEdit#TrackNameEdit:focus {
                border-bottom: 1px solid #888888;
                background-color: rgba(255, 255, 255, 0.03);
            }
            QPushButton#MuteButton, QPushButton#SoloButton, QPushButton#ArmButton {
                font-weight: bold;
                border-radius: 3px;
                min-width: 26px;
                min-height: 26px;
                max-width: 26px;
                color: #d4d4d4;
                background-color: #252526;
                border: 1px solid #3e3e42;
            }
            QPushButton#MuteButton:checked {
                background-color: #6b5317;
                color: #ffffff;
                border-color: #6b5317;
            }
            QPushButton#SoloButton:checked {
                background-color: #2b5a30;
                color: #ffffff;
                border-color: #2b5a30;
            }
            QPushButton#ArmButton:checked {
                background-color: #802b2b;
                color: #ffffff;
                border-color: #802b2b;
            }
            QComboBox#InputChannelCombo {
                background-color: #252526;
                border: 1px solid #3e3e42;
                border-radius: 3px;
                color: #d4d4d4;
                padding: 4px;
                font-size: 11px;
            }
            QComboBox#InputChannelCombo QAbstractItemView {
                background-color: #252526;
                color: #d4d4d4;
                selection-background-color: #4a4a4a;
                selection-color: #ffffff;
            }
            QPushButton#DeleteTrackButton {
                background-color: transparent;
                border: 1px solid rgba(166, 62, 62, 0.3);
                border-radius: 3px;
                color: #a63e3e;
                font-size: 10px;
                padding: 3px;
            }
            QPushButton#DeleteTrackButton:hover {
                background-color: rgba(166, 62, 62, 0.1);
                border-color: #a63e3e;
            }
            QPushButton#FxButton {
                background-color: #2d2d2d;
                border: 1px solid #3e3e42;
                border-radius: 3px;
                color: #e0e0e0;
                font-weight: bold;
                font-size: 11px;
                padding: 5px;
            }
            QPushButton#FxButton:hover {
                background-color: #444444;
                color: #ffffff;
                border-color: #555555;
            }
            QSlider#VolumeSlider::groove:vertical {
                background: #18181c;
                width: 6px;
                border-radius: 3px;
            }
            QSlider#VolumeSlider::sub-page:vertical {
                background: #888888;
                width: 6px;
                border-radius: 3px;
            }
            QSlider#VolumeSlider::handle:vertical {
                background: #505050;
                border: 1px solid #666;
                height: 12px;
                margin-left: -4px;
                margin-right: -4px;
                border-radius: 6px;
            }
            QSlider#VolumeSlider::handle:vertical:hover {
                background: #888888;
                border-color: #888888;
            }
            QLabel#VolDbLabel {
                color: #9d9d9d;
                font-size: 10px;
            }
        """)

    def populate_inputs(self):
        """Populates the input selector dropdown filtering by global device settings."""
        self.combo_input.clear()
        
        if self.audio_engine.enable_inputs:
            first_in = self.audio_engine.input_first_channel
            last_in = self.audio_engine.input_last_channel
            for i in range(first_in, last_in + 1):
                self.combo_input.addItem(f"Input Channel {i+1}", i)
                
        # Add Demo Loop option
        self.combo_input.addItem("Guitar Demo Loop", "loop")
        
        # Set current selection
        idx = self.combo_input.findData(self.track.input_channel)
        if idx >= 0:
            self.combo_input.setCurrentIndex(idx)
        else:
            self.combo_input.setCurrentIndex(self.combo_input.count() - 1)  # Default to loop if not found

    def update_levels(self):
        """Updates the LED VU level meter with the current peak dB."""
        self.level_meter.set_level(self.track.level_history)

    def set_selected(self, selected):
        """Sets selected state and updates outline glow."""
        self.is_selected = selected
        self.update_selection_style()
        if selected:
            self.trackSelected.emit(self.track)
            
    def update_selection_style(self):
        """Updates the styling based on selection state."""
        if self.is_selected:
            # Active selected card style (graphite highlight border)
            self.setStyleSheet(self.styleSheet() + """
                TrackCard {
                    border-left: 4px solid #888888;
                    background-color: #333333;
                }
            """)
        else:
            # Inactive card style
            self.setStyleSheet(self.styleSheet() + """
                TrackCard {
                    border-left: 4px solid transparent;
                    background-color: #2a2a2a;
                }
            """)

    def mousePressEvent(self, event):
        """Select track when clicking anywhere on the card."""
        self.set_selected(True)
        super().mousePressEvent(event)
        
    def on_rename(self):
        new_name = self.name_edit.text().strip()
        if new_name:
            self.track.name = new_name
            self.update()
            
    def on_mute_clicked(self):
        self.track.mute = self.btn_mute.isChecked()
        
    def on_solo_clicked(self):
        self.track.solo = self.btn_solo.isChecked()
        
    def on_arm_clicked(self):
        self.track.armed = self.btn_arm.isChecked()
        
    def on_input_changed(self):
        ch_data = self.combo_input.currentData()
        if ch_data is not None:
            self.track.input_channel = ch_data
            
    def on_volume_changed(self):
        val_db = self.vol_slider.value() / 10.0
        self.track.volume = val_db
        self.update_volume_label(val_db)
        
    def update_volume_label(self, val_db):
        if val_db <= -60.0:
            self.vol_label.setText("-inf dB")
        else:
            self.vol_label.setText(f"{val_db:+.1f} dB")
            
    def on_pan_changed(self, value):
        self.track.pan = value
        
    def on_fx_clicked(self):
        self.set_selected(True)
        # We can handle custom actions/notifications for fx rack focus
        
    def on_delete_clicked(self):
        self.trackRemoved.emit(self.track)

File: cpp/src/gui/TrackCard.cpp
#include "TrackCard.h"
#include "../audio/AudioEngine.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QMouseEvent>

namespace gui {

TrackCard::TrackCard(int trackIndex, const QString& trackName, dsp::AudioEngine* engine, QWidget* parent)
    : QWidget(parent), m_trackIndex(trackIndex), m_engine(engine)
{
    setFixedHeight(100);
    setObjectName("TrackCard");
    
    // Style specific to TrackCard
    setStyleSheet(R"(
        QWidget#TrackCard {
            background-color: #0b0b0c;
            border: 1px solid #222225;
            border-radius: 4px;
            margin-bottom: 4px;
        }
        QPushButton {
            min-width: 24px;
            padding: 2px 4px;
        }
        QPushButton:checked {
            background-color: #ff0033;
            border-color: #ff0033;
            color: #ffffff;
        }
        QSlider::groove:horizontal {
            border: 1px solid #222225;
            height: 4px;
            background: #000000;
        }
        QSlider::handle:horizontal {
            background: #88888c;
            width: 10px;
            margin: -4px 0;
            border-radius: 2px;
        }
        QSlider::handle:horizontal:hover {
            background: #ffffff;
        }
        QProgressBar {
            background-color: #1a1a1c;
            border: 1px solid #222225;
            border-radius: 2px;
            text-align: center;
        }
        QProgressBar::chunk {
            background-color: #00ff00;
        }
    )");

    QVBoxLayout* mainLayout = new QVBoxLayout(this);
    mainLayout->setContentsMargins(8, 8, 8, 8);
    mainLayout->setSpacing(6);

    // Header (Name + Mute/Solo/Arm)
    QHBoxLayout* headerLayout = new QHBoxLayout();
    m_nameLabel = new QLabel(trackName, this);
    m_nameLabel->setStyleSheet("font-weight: bold; color: #ffffff;");
    headerLayout->addWidget(m_nameLabel);
    
    headerLayout->addStretch();
    
    m_btnMute = new QPushButton("M", this);
    m_btnMute->setCheckable(true);
    m_btnSolo = new QPushButton("S", this);
    m_btnSolo->setCheckable(true);
    m_btnArm = new QPushButton(QChar(0x25CF), this); // Record circle
    m_btnArm->setCheckable(true);
    
    m_panDial = new CustomKnob(this);
    m_panDial->setRange(0, 100);
    m_panDial->setValue(50);
    m_panDial->setFixedSize(30, 30);
    m_panDial->setToolTip("Pan");
    
    headerLayout->addWidget(m_btnMute);
    headerLayout->addWidget(m_btnSolo);
    headerLayout->addWidget(m_btnArm);
    headerLayout->addWidget(m_panDial);
    
    mainLayout->addLayout(headerLayout);

    // No volume slider here anymore
    
    // Meters
    QVBoxLayout* meterLayout = new QVBoxLayout();
    meterLayout->setSpacing(2);
    m_meterL = new LevelMeter(Qt::Horizontal, this);
    m_meterL->setFixedHeight(6);
    
    m_meterR = new LevelMeter(Qt::Horizontal, this);
    m_meterR->setFixedHeight(6);
    
    meterLayout->addWidget(m_meterL);
    meterLayout->addWidget(m_meterR);
    mainLayout->addLayout(meterLayout);
    
    // Connect signals
    connect(m_panDial, &QDial::valueChanged, this, &TrackCard::onPanChanged);
    connect(m_btnMute, &QPushButton::toggled, this, &TrackCard::onMuteToggled);
    connect(m_btnSolo, &QPushButton::toggled, this, &TrackCard::onSoloToggled);
    connect(m_btnArm, &QPushButton::toggled, this, &TrackCard::onArmToggled);
    
    m_meterTimer = new QTimer(this);
    connect(m_meterTimer, &QTimer::timeout, this, &TrackCard::updateMeters);
    m_meterTimer->start(16);
}

void TrackCard::updateMeters()
{
    if (m_engine && m_engine->isEnginePlaying()) {
        float pL = m_engine->getTrackPeakL(m_trackIndex);
        float pR = m_engine->getTrackPeakR(m_trackIndex);
        
        m_meterL->setLevel(pL);
        m_meterR->setLevel(pR);
        
        // Synchronize pan state from engine
        float enginePan = m_engine->getTrackPan(m_trackIndex); // -1.0 to 1.0
        int panVal = static_cast<int>((enginePan * 50.0f) + 50.0f);
        m_panDial->blockSignals(true);
        m_panDial->setValue(panVal);
        m_panDial->blockSignals(false);
    } else {
        m_meterL->setLevel(0.0f);
        m_meterR->setLevel(0.0f);
    }
}

void TrackCard::mousePressEvent(QMouseEvent* event)
{
    emit clicked(m_trackIndex);
    event->ignore(); // allow QListWidget to process drag
}

void TrackCard::setSelected(bool selected)
{
    if (selected) {
        setStyleSheet(styleSheet() + " QWidget#TrackCard { border: 2px solid #00ff00; background-color: #1a1a1c; }");
    } else {
        setStyleSheet(styleSheet() + " QWidget#TrackCard { border: none; border-bottom: 1px solid #222225; background-color: #111111; }");
    }
}

void TrackCard::onMuteToggled(bool checked)
{
    if (m_engine) {
        dsp::EngineMessage msg;
        msg.type = dsp::EngineCommandType::SetTrackMute;
        msg.trackIndex = m_trackIndex;
        msg.boolValue = checked;
        m_engine->sendMessageFromUI(msg);
    }
}

void TrackCard::onSoloToggled(bool checked)
{
    if (m_engine) {
        dsp::EngineMessage msg;
        msg.type = dsp::EngineCommandType::SetTrackSolo;
        msg.trackIndex = m_trackIndex;
        msg.boolValue = checked;
        m_engine->sendMessageFromUI(msg);
    }
}

void TrackCard::onArmToggled(bool checked)
{
    if (m_engine) {
        dsp::EngineMessage msg;
        msg.type = dsp::EngineCommandType::SetTrackArm;
        msg.trackIndex = m_trackIndex;
        msg.boolValue = checked;
        m_engine->sendMessageFromUI(msg);
    }
}

void TrackCard::onPanChanged(int value)
{
    if (m_engine) {
        dsp::EngineMessage msg;
        msg.type = dsp::EngineCommandType::SetTrackPan;
        msg.trackIndex = m_trackIndex;
        msg.floatValue = (value - 50) / 50.0f; // Assuming 0-100 range, where 50 is center
        m_engine->sendMessageFromUI(msg);
    }
}

} // namespace gui
